### ISSUE TITLE
1.0.2 causes crash

### ISSUE THREAD

**Issue by bonebox** [hyperoslo/Sync#82]
I had Sync working with version 1.0.1, then updated to 1.0.2 today and now my app crashes with the below.  I reverted back to 1.0.1, and it works again.

```
*** Terminating app due to uncaught exception 'NSInvalidArgumentException', reason: '*** -[NSDictionary initWithObjects:forKeys:]: count of objects (4) differs from count of keys (0)'
*** First throw call stack:
(
	0   CoreFoundation                      0x0000000103907c65 __exceptionPreprocess + 165
	1   libobjc.A.dylib                     0x0000000105472bb7 objc_exception_throw + 45
	2   CoreFoundation                      0x0000000103850210 -[NSDate isEqual:] + 0
	3   CoreFoundation                      0x00000001038628e4 +[NSDictionary dictionaryWithObjects:forKeys:] + 52
	4   DATAFilter                          0x000000010337033e +[DATAFilter changes:inEntityNamed:localKey:remoteKey:context:predicate:inserted:updated:] + 702
	5   Sync                                0x00000001033a29cc +[Sync changes:inEntityNamed:predicate:parent:inContext:dataStack:completion:] + 1484
	6   Sync                                0x00000001033a3ef2 -[NSManagedObject(SyncPrivate) sync_processToManyRelationship:usingDictionary:andParent:dataStack:] + 1986
	7   Sync                                0x00000001033a335e -[NSManagedObject(SyncPrivate) sync_processRelationshipsUsingDictionary:andParent:dataStack:error:] + 558
	8   Sync                                0x00000001033a45b5 -[NSManagedObject(SyncPrivate) sync_processToOneRelationship:usingDictionary:andParent:dataStack:error:] + 1365
	9   Sync                                0x00000001033a35fd -[NSManagedObject(SyncPrivate) sync_processRelationshipsUsingDictionary:andParent:dataStack:error:] + 1229
	10  Sync                                0x00000001033a2cbc __78+[Sync changes:inEntityNamed:predicate:parent:inContext:dataStack:completion:]_block_invoke + 172
	11  DATAFilter                          0x0000000103370816 +[DATAFilter changes:inEntityNamed:localKey:remoteKey:context:predicate:inserted:updated:] + 1942
	12  Sync                                0x00000001033a29cc +[Sync changes:inEntityNamed:predicate:parent:inContext:dataStack:completion:] + 1484
	13  Sync                                0x00000001033a1c69 __61+[Sync changes:inEntityNamed:predicate:dataStack:completion:]_block_invoke + 121
	14  DATAStack                           0x000000010337d76a __43-[DATAStack performInNewBackgroundContext:]_block_invoke + 58
	15  CoreData                            0x0000000103469a79 developerSubmittedBlockToNSManagedObjectContextPerform + 201
	16  libdispatch.dylib                   0x0000000106939614 _dispatch_client_callout + 8
	17  libdispatch.dylib                   0x00000001069206a7 _dispatch_queue_drain + 2176
	18  libdispatch.dylib                   0x000000010691fcc0 _dispatch_queue_invoke + 235
	19  libdispatch.dylib                   0x00000001069233b9 _dispatch_root_queue_drain + 1359
	20  libdispatch.dylib                   0x0000000106924b17 _dispatch_worker_thread3 + 111
	21  libsystem_pthread.dylib             0x0000000106ca6637 _pthread_wqthread + 729
	22  libsystem_pthread.dylib             0x0000000106ca440d start_wqthread + 13
)
libc++abi.dylib: terminating with uncaught exception of type NSException
```

**Comment by 3lvis** [comment]
@bonebox Hey Jeremy! Damn, this sucks, could you provide me an example of your model and the JSON that you're using, please?

**Comment by bonebox** [comment]
I think I figured out where the problem is occurring, and it actually exists in 1.0.1 as well, but just doesn't cause a crash. I have an entity that that has a one-to-many relationship with itself.  After performing the sync, I noticed that the parent object gets populated ok, but none of the child objects do.  There are 4 of them, which I think is where the `count of objects (4) differs from count of keys (0)` is coming from.  

Here's the data you requested. Let me know if you need anything else.

```
...       
 "element" : {
            "type" : "group",
            "label" : "Triage",
            "elements" : [
                {
                    "type": "select",
                    "label": "Reason for Visit",
                    "reference": "reasonForVisit",
                    "selection": "open",
                    "items": [
                        { "value": "none", "label": "None" }
                    ]
                },
                {
                    "type": "select",
                    "label": "Allergies",
                    "reference": "allergies",
                    "selection": "open",
                    "items": [
                        { "value": "none/no known allergy", "label": "None/No Known Allergy" }
                    ]
                },
                {
                    "type": "select",
                    "label": "Last Vital Signs",
                    "reference": "lastVitalSigns",
                    "selection": "open",
                    "items": [
                        { "value": "none", "label": "None" }
                    ]
                },
                {
                    "type": "select",
                    "label": "Current Medications",
                    "reference": "currentMedications",
                    "selection": "open",
                    "items": [
                        { "value": "none", "label": "None" }
                    ]
                }
            ]
        }
...
```

![screen shot 2015-04-21 at 6 38 51 pm](https://cloud.githubusercontent.com/assets/1120504/7264801/cc6cdc16-e855-11e4-886f-aee53b4e4060.png)

**Comment by 3lvis** [comment]
Thanks! I'll make a test case to reproduce the issue.

Have a great day (or night🌚)!

**Comment by bonebox** [comment]
Well, I discovered the problem!

I have `hyper.isPrimaryKey` on the property `elementType` of the `Element` entity.  The corresponding JSON property is `type` (as you can see in my previous post), so I set it up this way according to the README since you have `type` listed as a reserved attribute.  Once I moved `hyper.isPrimaryKey` to a different property, it works!  So it seems there is a bug that has something to do with reserved attributes can't also be used as primary keys.

**Comment by 3lvis** [comment]
@bonebox We did a bunch of improvements for many-to-many relationships on [Sync 1.0.3](https://github.com/hyperoslo/Sync/releases/tag/1.0.3) would you mind verifying if this is still an issue? Thanks in advance :)

**Comment by bonebox** [comment]
Hey @3lvis sorry, forgot about this.  I just tested it, and it does work using a reserved type as primary key.  What I found though, is in addition to specifying the `isPrimaryKey` value, you *also* have to specify the `remoteKey` value, otherwise it will crash.  That may be worth clarifying in the documentation, as it could be confusing for people, since they normally wouldn't need to specify `remoteKey` (if the attribute is named appropriately).
